### PR TITLE
Fixed #27629 -- Fixed bug where allow_relation() is never called

### PR DESCRIPTION
**Pull request by srfilipek**
Modified the behavior surrounding router.allow_relation checks within
the related_descriptors module. It will now always resolve the
instance and value database if needed and then always call the
relation check.

https://code.djangoproject.com/ticket/27629

### CODE REVIEW

**Comment by srfilipek** [comment]
This does impact data consistency/correctness, so I hope to have it backported to at least 1.11.x.

File: tests/gis_tests/layermap/tests.py [hunk 1]
```diff
@@ -319,7 +319,8 @@ def db_for_write(self, model, **hints):
         return self.db_for_read(model, **hints)
 
     def allow_relation(self, obj1, obj2, **hints):
-        return None
+        # Required to be True after fixing bug #27629
+        return True
```
Comment (srfilipek) [inline tests/gis_tests/layermap/tests.py]: Because allow_relation is now called more often (correctly), I'm worried that improperly configured routers that were never fully exercised before may start to cause failures in existing systems that use Django.

File: tests/allow_relation/models.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+class MainModel(models.Model):
+    other = models.ForeignKey('ForeignBase', null=True, on_delete=models.CASCADE)
+
+
+class MainManyModel(models.Model):
+    others = models.ManyToManyField('ForeignBase')
+
+
+class MainOneOneModel(models.Model):
```
Comment (akulakov) [inline tests/allow_relation/models.py]: Should be `MainOneToOneModel` ?

Comment (srfilipek) [reply tests/allow_relation/models.py]: Sure, fixed.

File: tests/gis_tests/layermap/tests.py (hunk 1 above)
Comment (auvipy) [reply tests/gis_tests/layermap/tests.py]: may be adding some notes about that on release note?

File: tests/allow_relation/models.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
+from django.db import models
```
Comment (timgraham) [inline tests/allow_relation/models.py]: Please try to reuse existing models rather than adding a new test app. `multiple_databases` might be a good candidate.

File: tests/gis_tests/layermap/tests.py (hunk 1 above)
Comment (timgraham) [reply tests/gis_tests/layermap/tests.py]: The documentation says that, `allow_relation()` should "return True if a relation between obj1 and obj2 should be allowed, False if the relation should be prevented, or None if the router has no opinion." I didn't investigate but I can't logically see how a correct implementation would require router that previously had no opinion to now return True.

File: tests/allow_relation/tests.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        MainModel.objects.get_or_create(other=foreign_a)
+        MainModel.objects.all().delete()
+        MainModel.objects.update_or_create(other=foreign_a)
+        MainModel.objects.all().delete()
+        ma = MainModel(other=foreign_a)
+        ma.save()
+        MainModel.objects.all().delete()
+
+        # MainModel -> ForeignB is not allowed
+        self.assertRaises(
```
Comment (timgraham) [inline tests/allow_relation/tests.py]: Use `assertRaisesMessage()`

File: tests/gis_tests/layermap/tests.py (hunk 1 above)
Comment (srfilipek) [reply tests/gis_tests/layermap/tests.py]: The problem is that the documentation is not clear what should happen if the router returns `None`. Are relationships by default allowed or not allowed?

The current and "fixed" implementation checks if a relationship is allowed via `if not router.allow_relation(...)`. See `related_descriptors.py` line 211 and 457 (old line numbers). A return value of `None` would mean that a relationship is not allowed (i.e., it must explicitly be allowed by returning `True`). I kept this behavior.

However, since `allow_relation()` was not being called **at all** in the specific scenario here (hence my pull request), this issue was never caught. If the proper conditions were met such that `allow_relation()` were to be called (such as re-assigning the foreign key after the initial save), then the relationship would never have been allowed in the first place.

Comment (srfilipek) [reply tests/gis_tests/layermap/tests.py]: Actually, `None` never gets returned to the code in `related_descriptors.py`...

The [current router behavior (utils.py)](https://github.com/django/django/blob/84dcd1624784c2239f96a97b3151145a85dfbbe3/django/db/utils.py#L281), if all the routers return None, is to disallow cross-database relationships.

The problem with this test is that the `allow_relation()` function here was never called, until now.

File: tests/allow_relation/models.py (hunk 3 above)
Comment (srfilipek) [reply tests/allow_relation/models.py]: I'll simplify the tests and get them working in that test app.

**Comment by srfilipek** [comment]
I'm trying to figure out why the mysql_gis tests are failing, but it's a little difficult given that I don't have a system in place to test it (is there a docker container somewhere?).

It appears that there may be other bugs getting in the way. There are objects being created during the fixture teardown post-migrate signal after a flush (why?), as the stack trace would have me believe, and apparently not all of them abide by the router in the test case (everything should have the database 'other', but some model instances are on database 'default').

Other than this one issue, I think the merge request has been updated with all the feedback I've received. The comment regarding improperly configured routers was put in the release notes for 2.0.5 and some clarification added to the `allow_relation()` function docs.

File: tests/gis_tests/layermap/tests.py [hunk 5]
```diff
@@ -319,7 +319,10 @@ def db_for_write(self, model, **hints):
         return self.db_for_read(model, **hints)
 
     def allow_relation(self, obj1, obj2, **hints):
-        return None
+        # Required after fixing issue #27629
+        # ContentType objects are created during a post-migrate signal while performing fixture teardown
+        # using the default database alias and do not abide by the database specified by this router.
+        return True
```
Comment (srfilipek) [inline tests/gis_tests/layermap/tests.py]: After some digging, it appears that `django/contrib/auth/management/__init__.py` `create_permissions()` is creating some `ContentType` objects using the default database alias (by specifying it to the db_manager) and not 'other' specified by the router. Some related objects are being created in the 'other' database, and the relation check is then failing since they exist in different databases.

I would argue the proper fix is to make the `create_permissions()` function use the routers properly. Short of that, this seems like the least invasive solution.

But, this still makes me worry that other systems in the wild may run into this hidden bug once this fix is applied.

File: docs/topics/db/multi-db.txt [hunk 6]
```diff
@@ -159,9 +159,10 @@ A database Router is a class that provides up to four methods:
 
     Return ``True`` if a relation between ``obj1`` and ``obj2`` should be
     allowed, ``False`` if the relation should be prevented, or ``None`` if
-    the router has no opinion. This is purely a validation operation,
-    used by foreign key and many to many operations to determine if a
-    relation should be allowed between two objects.
+    the router has no opinion. If no router has an opinion, the default
```
Comment (timgraham) [inline docs/topics/db/multi-db.txt]: This documentation change isn't related to the bug fix, correct? Please submit a separate pull request.

File: docs/releases/2.0.5.txt [hunk 7]
```diff
@@ -23,3 +23,8 @@ Bugfixes
   ``values_list()`` after combining an annotated and unannotated queryset with
   ``union()``, ``difference()``, or ``intersection()`` crashed due to mismatching
   columns (:ticket:`29286`).
+
+* Fixed issue where the ``allow_relation()`` check was never enforced.
+  Improperly configured routers may have gone unnoticed and may need to be
+  updated. Note that inter-database relationships require a router to return
```
Comment (timgraham) [inline docs/releases/2.0.5.txt]: inter-database -> cross-database

File: docs/releases/2.0.5.txt [hunk 8]
```diff
@@ -23,3 +23,8 @@ Bugfixes
   ``values_list()`` after combining an annotated and unannotated queryset with
   ``union()``, ``difference()``, or ``intersection()`` crashed due to mismatching
   columns (:ticket:`29286`).
+
+* Fixed issue where the ``allow_relation()`` check was never enforced.
```
Comment (timgraham) [inline docs/releases/2.0.5.txt]: I don't believe this qualifies for a backport per our [supported versions policy](https://docs.djangoproject.com/en/dev/internals/release-process/#supported-versions). Especially given the possibility of breaking existing code. Let's target this for Django 2.1 -- a note can be added in the "backwards incompatible changes" section of the release notes.

File: tests/multiple_database/models.py [hunk 9]
```diff
@@ -80,3 +80,11 @@ class UserProfile(models.Model):
 
     class Meta:
         ordering = ('flavor',)
+
+
+class ForeignRelation(models.Model):
```
Comment (timgraham) [inline tests/multiple_database/models.py]: I don't think a new model is needed. There are existing models that have ForeignKey/ManyToManyField/OneToOneField that could be used. Also, there are a number of existing tests (e.g. `test_(foreign/o2o/m2m)_key_cross_database_protection` that may duplicate some of the added tests). In particular, any tests that pass prior to this fix should be added separately.

File: tests/gis_tests/layermap/tests.py (hunk 5 above)
Comment (timgraham) [reply tests/gis_tests/layermap/tests.py]: I'll investigate this later.

File: tests/multiple_database/tests.py [hunk 10]
```diff
@@ -2082,3 +2082,120 @@ def test_reverse_m2m_update(self):
...
+    """
+    Router for testing that relationships between foreign objects are enforced
+    """
+    def allow_relation(self, obj1, obj2, **hints):
+        # Cause all foreign relations to fail
+        return False
+
+
+@override_settings(DATABASE_ROUTERS=[NoRelationRouter()])
+class AllowRelationTests(TestCase):
```
Comment (srfilipek) [inline tests/multiple_database/tests.py]: This has been reduced to be only test cases that fail prior to this fix.

Comment (timgraham) [reply tests/multiple_database/tests.py]: Still, I wonder if so many tests are needed. I think it would be enough to have one test for ForwardManyToOneDescriptor and ReverseOneToOneDescriptor. In other words, every place that the descriptor is called doesn't need to be tested. Testing the each descriptor once is enough, I think.

Please use the context manager version of assertRaisesMessage and follow the style of similar assertions in this file.

Comment (srfilipek) [reply tests/multiple_database/tests.py]: The multitude of constructor-variant tests could be removed, but the messed up logic that used to surround `allow_relation()` results in different behavior based on the state of the instance or value objects (whether they are associated with a database or not at the time of the check).

Ensuring those test cases are covered is needed, but that also involves multiple tests per descriptor type.

Comment (srfilipek) [reply tests/multiple_database/tests.py]: If the testing strategy is to ensure the new implementation is properly tested (not necessarily the all conditions of the old implementation), then I agree that one test per descriptor is sufficient.

I've updated the tests further. Now there are only two.
